bin/consequence_mapping/consequence_mapping.py

```diff
@@ -0,0 +1,154 @@
+        consequences.sort(key=lambda consequence: abs(consequence.get('distance', 0)))
+
+        # If mandated by a flag, keep only one (least distant) consequence
+        if only_closest:
+            consequences = [consequences[0]]
+
+        # Return a subset of fields (required for output) of filtered consequences
+        results_by_variant[variant_identifier].extend([
+            (variant_identifier, c['gene_id'], c['gene_symbol'], most_severe_consequence_term, c.get('distance', 0))
+            for c in consequences
+        ])
+
+
+def get_variants_without_consequences(results_by_variant):
+    """Returns a list of variants for which no consequences were found."""
+    return sorted({
+        variant_id
+        for variant_id, list_of_consequences in results_by_variant.items()
+        if len(list_of_consequences) == 0
+    })
+
+
+def process_variants(variants):
+    """Given a list of variant IDs, return a list of consequence types (each including Ensembl gene name & ID and a
+    functional consequence code) for a given variant."""
+
+    # First, we query VEP with default parameters, looking for variants affecting protein coding and miRNA transcripts
+    # up to a standard distance (5000 nucleotides either way, which is default for VEP) from the variant.
+    results_by_variant = {}
+    vep_results = query_vep(variants=variants, search_distance=5000)
+    extract_consequences(vep_results=vep_results, acceptable_biotypes={'protein_coding', 'miRNA'},
+                         only_closest=False, results_by_variant=results_by_variant)
+
+    # See if there are variants with no consequences up to the default distance
+    variants_without_consequences = get_variants_without_consequences(results_by_variant)
+    # If there are, we will now do a second round of querying, this time looking only at protein coding biotypes (vs.
+    # miRNA *and* protein coding during the first round) up to a distance of 500,000 bases each way.
+    if variants_without_consequences:
+        logger.info('Found {} variant(s) without standard consequences: {}. Querying distant regions'.format(
+            len(variants_without_consequences), '|'.join(variants_without_consequences)))
+        distant_vep_results = query_vep(variants=variants_without_consequences, search_distance=500000)
+        extract_consequences(vep_results=distant_vep_results, acceptable_biotypes={'protein_coding'},
+                             only_closest=True, results_by_variant=results_by_variant)
+
+    # See if there are still variants with no consequences, even up to a wide search window
+    variants_without_consequences = get_variants_without_consequences(results_by_variant)
+    if variants_without_consequences:
+        logger.info('After distant querying, still remaining {} variant(s) without consequences: {}'.format(
+            len(variants_without_consequences), '|'.join(variants_without_consequences)
```

Comment:
Suggest newline instead of pipe for readability...

Reply:
Yes, exactly, it's there for easier programmatic processing of the logs. I can't replace it with newline because it would break `logging` module prefixes, although I could replace it with some other symbol or combination of symbols.

Reply:
On second thought, pipe looks fine... you can use the cut command with the pipe delimiter in conjunction with grep to get the list of variants...